### ISSUE TITLE
When viewing recent builds error is logged: exceeded memory limit for $group, but didn't allow external sort

### ISSUE THREAD

**Issue by t00** [sorry-cypress/sorry-cypress#753]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/sorry-cypress/sorry-cypress/issues?q=is%3Aopen+is%3Aissue) and [discussions](https://github.com/sorry-cypress/sorry-cypress/discussions).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have read the [documentation](https://docs.sorry-cypress.dev/) and found no answer to my problem

### Environment information

<details>
```
  - sorry-cypress version: `2.5.1`
  - platform: `docker`
  - service: `api`
```
</details>

### Describe the bug

We have 297 builds with over 700 tests each with 2 repeats (up to 3 runs total) and it seems mongo cannot cope with a group query taking more than 100MB RAM when viewing "Recent Builds" section, which in fact I was never able to view as it was added when we already had collected a lot of builds.

The UI shows: Error loading data
The api log shows:

`
2023-02-25T10:27:43.884+0000 I  COMMAND  [conn16] command sorry-cypress.runs command: aggregate { aggregate: "runs", pipeline: [ { $match: {} }, { $sort: { _id: -1 } }, { $limit: 1000 }, { $group: { _id: "$meta.ciBuildId", runs: { $push: "$$ROOT" }, runId: { $min: "$_id" } } }, { $sort: { runId: -1 } }, { $limit: 20 }, { $addFields: { ciBuildId: "$_id", createdAt: { $arrayElemAt: [ "$runs.createdAt", -1 ] }, updatedAt: { $max: "$runs.progress.updatedAt" } } } ], cursor: {}, lsid: { id: UUID("12345678-1234-1234-1234-123456781234") }, $db: "sorry-cypress" } planSummary: IXSCAN { _id: 1 } numYields:17 queryHash:96B27E33 planCacheKey:96B27E33 ok:0 errMsg:"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in." errName:Location16945 errCode:16945 reslen:188 locks:{ ReplicationStateTransition: { acquireCount: { w: 39 } }, Global: { acquireCount: { r: 39 } }, Database: { acquireCount: { r: 39 } }, Collection: { acquireCount: { r: 39 } }, Mutex: { acquireCount: { r: 22 } } } storage:{} protocol:op_msg 301ms
`

It seems the error is self-resolving, adding `allowDiskUse` to the respective Recent Builds aggregating query should fix it.

### Expected behavior

Recent Builds shown

### Reproduction steps

Collect a lot of test runs and open Recent Builds.

### Full log and debug output

<details>

```
2023-02-25T10:27:33.211+0000 I  COMMAND  [conn19] command sorry-cypress.runs command: aggregate { aggregate: "runs", pipeline: [ { $match: {} }, { $sort: { _id: -1 } }, { $limit: 1000 }, { $group: { _id: "$meta.ciBuildId", runs: { $push: "$$ROOT" }, runId: { $min: "$_id" } } }, { $sort: { runId: -1 } }, { $limit: 20 }, { $addFields: { ciBuildId: "$_id", createdAt: { $arrayElemAt: [ "$runs.createdAt", -1 ] }, updatedAt: { $max: "$runs.progress.updatedAt" } } } ], cursor: {}, lsid: { id: UUID("12345678-1234-1234-1234-123456781234") }, $db: "sorry-cypress" } planSummary: IXSCAN { _id: 1 } numYields:16 queryHash:96B27E33 planCacheKey:96B27E33 ok:0 errMsg:"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in." errName:Location16945 errCode:16945 reslen:188 locks:{ ReplicationStateTransition: { acquireCount: { w: 38 } }, Global: { acquireCount: { r: 38 } }, Database: { acquireCount: { r: 38 } }, Collection: { acquireCount: { r: 38 } }, Mutex: { acquireCount: { r: 22 } } } storage:{} protocol:op_msg 315ms
2023-02-25T10:27:38.545+0000 I  COMMAND  [conn3] command sorry-cypress.runs command: aggregate { aggregate: "runs", pipeline: [ { $match: {} }, { $sort: { _id: -1 } }, { $limit: 1000 }, { $group: { _id: "$meta.ciBuildId", runs: { $push: "$$ROOT" }, runId: { $min: "$_id" } } }, { $sort: { runId: -1 } }, { $limit: 20 }, { $addFields: { ciBuildId: "$_id", createdAt: { $arrayElemAt: [ "$runs.createdAt", -1 ] }, updatedAt: { $max: "$runs.progress.updatedAt" } } } ], cursor: {}, lsid: { id: UUID("12345678-1234-1234-1234-123456781234") }, $db: "sorry-cypress" } planSummary: IXSCAN { _id: 1 } numYields:16 queryHash:96B27E33 planCacheKey:96B27E33 ok:0 errMsg:"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in." errName:Location16945 errCode:16945 reslen:188 locks:{ ReplicationStateTransition: { acquireCount: { w: 38 } }, Global: { acquireCount: { r: 38 } }, Database: { acquireCount: { r: 38 } }, Collection: { acquireCount: { r: 38 } }, Mutex: { acquireCount: { r: 22 } } } storage:{} protocol:op_msg 302ms
2023-02-25T10:27:43.884+0000 I  COMMAND  [conn16] command sorry-cypress.runs command: aggregate { aggregate: "runs", pipeline: [ { $match: {} }, { $sort: { _id: -1 } }, { $limit: 1000 }, { $group: { _id: "$meta.ciBuildId", runs: { $push: "$$ROOT" }, runId: { $min: "$_id" } } }, { $sort: { runId: -1 } }, { $limit: 20 }, { $addFields: { ciBuildId: "$_id", createdAt: { $arrayElemAt: [ "$runs.createdAt", -1 ] }, updatedAt: { $max: "$runs.progress.updatedAt" } } } ], cursor: {}, lsid: { id: UUID("12345678-1234-1234-1234-123456781234") }, $db: "sorry-cypress" } planSummary: IXSCAN { _id: 1 } numYields:17 queryHash:96B27E33 planCacheKey:96B27E33 ok:0 errMsg:"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in." errName:Location16945 errCode:16945 reslen:188 locks:{ ReplicationStateTransition: { acquireCount: { w: 39 } }, Global: { acquireCount: { r: 39 } }, Database: { acquireCount: { r: 39 } }, Collection: { acquireCount: { r: 39 } }, Mutex: { acquireCount: { r: 22 } } } storage:{} protocol:op_msg 301ms
```

</details>

**Comment by mathpaquette** [comment]
@t00 okay let me try to reproduce as well on my side.

**Comment by mathpaquette** [comment]
@t00 adding allowDiskUse might introduce other issues as mentioned here: https://stackoverflow.com/a/57541306/2228237

If you only have 1-2 projects, you can tweak this environment parameter to decrease the batch size of grouping projects by unique build id. 

https://github.com/sorry-cypress/sorry-cypress/blob/ad9e86414c903bd80ec26d2e4740f6901a47bf0e/packages/api/src/config.ts#L6
https://github.com/sorry-cypress/sorry-cypress/blame/master/packages/api/src/datasources/runs.ts#L97

The maths behind are the following. Basically we need a safety buffer to make sure that no runs have been left behind during grouping operation.

`CI_BUILD_BATCH_SIZE = 1000`
`PAGE_ITEMS_LIMIT = 20`

hypothetical number of projects = 25

CI_BUILD_BATCH_SIZE / nb_projects = 40 groups
40 groups / PAGE_ITEMS_LIMIT = 2 = safety buffer

Basically, runs can be unorderly added in the database if you have multiple builds running so its important to have a big enough safety buffer.

Anyways, try to target a safety buffer of 2 depending of the number of projects you have.

This CI_BUILD_BATCH_SIZE can be set as a env variables while creating the api container.

Let me know.

**Comment by mathpaquette** [comment]
@agoldis most probably, I would expect that teams have less than 5 projects in general ?

**Comment by mathpaquette** [comment]
@t00 any updates ?

**Comment by t00** [comment]
I tried back then and now again and was unable to change the outcome by tweaking CI_BUILD_BATCH_SIZE and PAGE_ITEMS_LIMIT environment variables.

Example docker-compose api part:

```
  api:
    image: agoldis/sorry-cypress-api:latest
    restart: always
    environment:
      MONGODB_URI: 'mongodb://mongo:27017'
      MONGODB_DATABASE: 'sorry-cypress'
      CI_BUILD_BATCH_SIZE: 5000
      PAGE_ITEMS_LIMIT: 100
    ports:
      - 4000:4000
    depends_on:
      - mongo

```

In all honestly, the problem would be best handled automatically as grouping can eventually hit any limit of builds. Possibly setting a very high maximum limits would help?

**Comment by mathpaquette** [comment]
Okay let me recheck that

**Comment by mathpaquette** [comment]
I'll change the defaults

**Comment by t00** [comment]
That would be great, I will test the docker image as soon as it is released.

**Comment by mathpaquette** [comment]
ill check dont worry mister bot.

**Comment by t00** [comment]
Just an update - v2.5.8 still has the problem:

2023-08-07T06:09:09.726+0000 I  COMMAND  [conn16] command sorry-cypress.runs command: aggregate { aggregate: "runs", pipeline: [ { $match: {} }, { $sort: { _id: -1 } }, { $limit: 1000 }, { $group: { _id: "$meta.ciBuildId", runs: { $push: "$$ROOT" }, runId: { $min: "$_id" } } }, { $sort: { runId: -1 } }, { $limit: 20 }, { $addFields: { ciBuildId: "$_id", createdAt: { $arrayElemAt: [ "$runs.createdAt", -1 ] }, updatedAt: { $max: "$runs.progress.updatedAt" } } } ], cursor: {}, lsid: { id: UUID("59a1d319-089f-4744-8dd6-6841a595f688") }, $db: "sorry-cypress" } planSummary: IXSCAN { _id: 1 } numYields:15 queryHash:96B27E33 planCacheKey:96B27E33 ok:0 errMsg:"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in." errName:Location16945 errCode:16945 reslen:188 locks:{ ReplicationStateTransition: { acquireCount: { w: 36 } }, Global: { acquireCount: { r: 36 } }, Database: { acquireCount: { r: 36 } }, Collection: { acquireCount: { r: 36 } }, Mutex: { acquireCount: { r: 21 } } } storage:{} protocol:op_msg 267ms
